BSP/STM32F413/Src/BSP_CAN.c

```diff
@@ -0,0 +1,377 @@
+/* Copyright (c) 2020 UT Longhorn Racing Solar */
+
+#include "BSP_CAN.h"
+#include "stm32f4xx.h"
+#include "os.h"
+
+// The message information that we care to receive
+typedef struct _msg {
+    uint32_t id;
+    uint8_t data[8];
+} msg_t;
+
+// Set up a fifo for receiving
+#define FIFO_TYPE msg_t
+#define FIFO_SIZE 10
+#define FIFO_NAME msg_queue
+#include "fifo.h"
+
+static msg_queue_t gRxQueue;
+
+// Required for receiving CAN messages
+static CanTxMsg gTxMessage;
+static CanRxMsg gRxMessage;
+
+// User parameters for CAN events
+static void (*gRxEvent)(void);
+static void (*gTxEnd)(void);
+
+
+void BSP_CAN1_Init();
+void BSP_CAN3_Init();
+
+/**
+ * @brief   Initializes the CAN module that communicates with the rest of the electrical system.
+ * @param   rxEvent : the function to execute when recieving a message. NULL for no action.
+ * @param   txEnd   : the function to execute after transmitting a message. NULL for no action.
+ * @return  None
+ */
+
+void BSP_CAN_Init(CAN_t bus, void (*rxEvent)(void), void (*txEnd)(void)) {
+
+    // Configure event handles
+    gRxEvent  = rxEvent;
+    gTxEnd    = txEnd;
```

Comment:
This initialization function will be called multiple times: one for each bus. Therefore, we should have multiple callbacks (one for each bus) or they will be overwritten.